### PR TITLE
Fix project config and package deploy to support the windows code in a different location

### PR DESCRIPTION
**Pull request by dannyvv**
There were 2 problems:

###autolink
Auto-Linking was picking up the wrong path, when a custom `react-native.config.js` configures the windows folder with the users code in a different location.
This PR ensures that the custom sourcePath property is copied to the final configuration so it works.

###app deployment
App deployment hardcoded in a bunch of places to optoins.root\windows. This change updates the code to pass the windowsRoot which is where the native app code lives.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6232)

### CODE REVIEW

**Comment by dannyvv** [comment]
@NickGerleman: This might become an ask to backport to 0.63. I'll hear tomorrow from customer.

**Review by NickGerleman — APPROVED** [review]
Lgtm, but going to turn off AutoMerge for a moment so the authors of autolinking/deploy code have a chance to take a look.

**Comment by dannyvv** [comment]
Added request to backport.

**Comment by asklar** [comment]
Discussed this with Danny offline. Would love to get Jon's take on it too.
My main concerns:
- does this break any react-native config scenarios? [sounds like it's ok but should verify]
- does this break any --sln overrides? [same]
- does it produce any difference in the package paths? we've seen that these paths are somewhat unstable across VS releases, and they're different between C++ and C# so whatever we can do to make finding the appx simpler and more deterministic, the better.

We have not explored what sharing one native app's scaffolding across multiple JS means... what does it mean for app.js, how does metro bundle, app1, app2 or which?
My 2c is we should steer customers towards the solution we have tried and fully thought through / tested and not create a new path or add special cases for one specific customer's benefit.

**Comment by NickGerleman** [comment]
I kind of have similar sentiments re the `--sln` option. I think it got added because `playground-win32` was added as a custom project copied/pasted from Playground instead of as a new app template, it broke `run-windows`, and we decided to add flexibility to `run-windows` to accommodate it instead of putting `playground-win32` in its own package.

OTOH if this is a strict improvement in correctness of what we have right now, it seems fine to add. I kind of like the idea of doing a cleanup of some of the CLI options eventually though. We've been bad enough at making sure the happy-path works correctly, that I think being more opinionated to lower complexity could make sense.

**Comment by jonthysell** [comment]
@dannyvv Can you provide an example folder structure where react-native config wasn't working?

File: packages/@react-native-windows/cli/src/config/projectConfig.ts [hunk 1]
```diff
@@ -99,6 +99,9 @@ export function projectConfigWindows(
   var validProject = false;
 
   if (usingManualOverride) {
```
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/config/projectConfig.ts:101]: Why is this necessary? This blanket overwrites the logic for the last several lines of code.

File: packages/@react-native-windows/cli/src/runWindows/utils/deploy.ts [hunk 2]
```diff
@@ -231,9 +239,17 @@ export async function deployToDesktop(
     windowsConfig && windowsConfig.project && windowsConfig.project.projectName
       ? windowsConfig.project.projectName
       : options.proj!;
-  const appPackageFolder = getAppPackage(options, projectName);
+  const windowsRoot =
```
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/runWindows/utils/deploy.ts:242]: Is there a situation where windowsConfig doesn't have a folder/sourceDir?

If windowsConfig === null, then we need to error and exit before this point. Developers are allowed to set the value to null to block an platform from being detected.

Otherwise, it looks like windowsRoot is basically just the absolute path for sourceDir - and any logic for sourceDir should be calculated in config, for everyone's benefit, not here.

File: packages/@react-native-windows/cli/src/config/projectConfig.ts (hunk 1 above)
Comment (jonthysell) [reply packages/@react-native-windows/cli/src/config/projectConfig.ts:101]: I see the problem, it's up in line 96 - I remember writing that, that's bad logic on my part.

that line should actually calculate the relative path from folder to sourceDir, then you won't need to set it again later.

Comment (dannyvv) [reply packages/@react-native-windows/cli/src/config/projectConfig.ts:101]: Thanks Jon, fixed it.

File: packages/@react-native-windows/cli/src/runWindows/utils/deploy.ts (hunk 2 above)
Comment (dannyvv) [reply packages/@react-native-windows/cli/src/runWindows/utils/deploy.ts:242]: I updated the code to fall back in this case to the slnFile. This should also take care of the case where folks pass -slnFile to provide a fallback from the config override.

**Comment by Khalef1** [comment]
Removing backport request as this pr was reverted.
